### PR TITLE
Non-Record: 8L + BigramHash(12288) + Systematic HyperOpt (val_bpb=1.2392, 1xH100, 129 experiments)

### PR DESCRIPTION
**Pull request by CrimsonSithria**
## Summary

Systematic hyperparameter optimization across **129 experiments** ($19.47 total compute) on 1x H100 SXM, achieving **val_bpb=1.2392** with a 15.9MB artifact.

## Key Contributions

1. **Hyperparameter Scaling Laws**: Mapped how LR, WD, batch size, and model depth must co-scale with GPU throughput and step count across 129 experiments
2. **BigramHash with Dimensionality Projection**: dim=128 projection saves 3.7MB artifact space with <0.002 BPB quality loss, enabling deeper models within 16MB
3. **Weight Decay as Compression Knob**: WD linearly controls int8+zlib artifact size (WD=0.03→17MB, WD=0.05→15.7MB for 8L model)
4. **Batch Size Scaling on H100**: 131K tokens (2x default) gives -0.027 BPB vs 65K batch

## Architecture

- 8 transformer layers, 512-dim, 8 heads (4 KV heads, GQA)
- 4x MLP with relu^2, BigramHash 12288 buckets with dim=128 projection
- RoPE base=50000, logit softcap=30
- Muon optimizer: LR=0.03, WD=0.048, batch=131K tokens

## Improvement Trajectory

| Stage | val_bpb |
|-------|---------|
| Baseline (9L) | 1.4691 |
| 5L + 4x MLP | 1.3997 |
| BigramHash + RoPE | 1.3873 |
| H100 + 7L + 131K batch | 1.2417 |
| **H100 + 8L + BIGRAM 12288** | **1.2392** |

## Comprehensive Negative Results (17 techniques tested and rejected)

SmearGate, EMA, SWA (<7k steps), orthogonal init, label smoothing, magnitude pruning, zstd-22, all alt activations (leaky_relu^2, x*abs(x), mish^2), 96K batch, BIGRAM_DIM=64, 9L 3x MLP -- all documented with specific BPB values.

## Run Command

```bash
NUM_LAYERS=8 MLP_MULT=4 TRAIN_SEQ_LEN=2048 EVAL_SEQ_LEN=2048 \
EVAL_STRIDE=512 TRAIN_BATCH_TOKENS=131072 GRAD_ACCUM_STEPS=1 \
MAX_WALLCLOCK_SECONDS=600 WARMDOWN_ITERS=3000 \
MATRIX_LR=0.03 SCALAR_LR=0.03 TIED_EMBED_LR=0.08 \
MUON_WD=0.048 GRAD_CLIP_NORM=0.3 QK_GAIN_INIT=1.0 \
BIGRAM_BUCKETS=12288 BIGRAM_DIM=128 ROPE_BASE=50000 \
LOGIT_SOFTCAP=30 MUON_BACKEND_STEPS=5 \
python train_gpt.py
```

**Hardware**: 1x H100 SXM 80GB, 10 min, ~134ms/step, 4486 steps

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-Record: 8L + BigramHash(12288) + Systematic HyperOpt (val_bpb=1.2392, 1xH100, 129 experiments)

**BPB:** 1.2392 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `b57868f96947`, file `records/track_non_record_16mb/2026-03-22_1xH100_Systematic_HyperOpt_129exp/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.05s, dim=512, layers=10, vocab=1024, code=60115 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.05s, dim=512, layers=10, vocab=1024, code=60115 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
